### ISSUE TITLE
Program doesn't exit when there are services 

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-lang#18601]
**Description:**
$title. According to the spec ("Module and program execution" and "Listeners and services"), the program should not exit only when there are listeners.

**Steps to reproduce:**
```ballerina
import ballerina/io;

public function main() {
    io:println("main");
}

service foo = service {
    resource function bar() {
        
    }
};
```

**Affected Versions:**
```cmd
$ ballerina -v
Ballerina 1.0.0-beta
Language specification 2019R3
Ballerina tool 1.0.0
```

**Comment by riyafa** [comment]
There are a few things that has to be decided on this issue:
(Note that spec reference in the following is from the sections `Module and program execution` and `Listeners and services`

1. Currently jar file is generated only if there's an entry point, currently an entry point is considered as a main method or a service initialization. According to spec module initialization is performed at all times(regardless of the existence of a main function). Hence it is good to generate the jar file at all times. If we decide against generating jar files at all times then we need to reconsider whether the entry point need to be considered as the service or the listener declaration. 
2. If there are listeners in the source code then it can exist as a listener declaration or as simply an object initialization. Only if the listener is as a listener declaration the lifecycle methods are called automatically. As I understand only if there's a listener declaration that the program should not exit. If there's an object initialization expression it is possible that the lifecycle methods are later called manually, but it may not be the case. Hence it is more correct to not terminate the program only for listener declaration.

**Comment by MaryamZi** [comment]
1. Yeah, AFAIU, the "Module and program execution" section of the spec doesn't require a module-listener-declaration or a main function to execute a program. Given that we can now have a user-defined module `init` function also, it may make sense to revisit this restriction. 

2. Yeah, given that `start` may or may not be called for such a listener object (that's not part of a listener-decl), I also think we should not terminate only for listener-decls. However, on a related note, I think there are two other not-so-common use-cases:
- creating a listener object and passing it as an argument to a function in another module which attaches a service to this listener and calls `start` (we do something quite similar for `WebSub`) or 
- creating a service and passing it as an argument to a function in another module which attaches it to a listener internally

   Given that we have such scenarios, I wonder if we should provide some construct that allows the user to explicitly say the program should not terminate. Right now, what we seem to be doing is introducing a dummy service variable or using `runtime:sleep` to keep the program from terminating.

**Comment by manuranga** [comment]
reopen because fix has been revered in #26532 due to #26528
